### PR TITLE
Sketch of C# support

### PR DESCRIPTION
**Pull request by Porges**
Here's a sketch of supporting LINQ for `Gen` and `Property` in C#.

What I've done here:

1. Add a test that all Gen/Property/Range combinators have standard naming (just tests for capital letter at the moment). Some of the names I've chosen are a little different to the F# ones (I'll summarize these later).
2. Add LINQ methods 
3. Add tests using those methods in C#

I'm not that happy with the `ForAll` needed to wrap all `Gens` in C# at the moment.

It might be better that all `Property`s are written direct-style, like:

    Property.ForAll(gen1, x => ... )

Rather than with LINQ syntax, and reserve the LINQ syntax for writing your own `Gen`s.

Related to https://github.com/hedgehogqa/fsharp-hedgehog/issues/88.

### CODE REVIEW

**Comment by Porges** [comment]
Will need to fix CI build to work with new-style .csproj ...

File: tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (moodmosaic) [inline tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests.csproj]: This project won't load on Visual Studio 2015. And there's #108 regarding Visual Studio 2017 😕

File: src/Hedgehog/Gen.fs [hunk 2]
```diff
@@ -221,6 +252,7 @@ module Gen =
     /// is halved. When the size gets to one or less, selections are no longer made
     /// from the recursive list.
     /// <i>The first argument (i.e. the non-recursive input list) must be non-empty.</i>
+    [<CompiledName("ChoiceRecursive")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Here we're deviating a bit, but a think it's okay (?) 😃 /cc @jystic

File: src/Hedgehog/Gen.fs [hunk 3]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
 
+    [<CompiledName("Char")>]
     // Generates a random character in the specified range.
     let char (lo : char) (hi : char) : Gen<char> =
         integral <| Range.constant (int lo) (int hi) |> map char
 
     /// Generates a Unicode character, including invalid standalone surrogates:
     /// '\000'..'\65535'
+    [<CompiledName("CharAll")>]
     let unicodeAll : Gen<char> =
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: I think this could be `UnicodeAll` (?)

File: src/Hedgehog/Gen.fs [hunk 4]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
         let lo = System.Char.MinValue
         let hi = System.Char.MaxValue
         char lo hi
 
     // Generates a random digit.
+    [<CompiledName("CharDigit")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 5]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let digit : Gen<char> =
         char '0' '9'
 
     // Generates a random lowercase character.
+    [<CompiledName("CharLower")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 6]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let lower : Gen<char> =
         char 'a' 'z'
 
     // Generates a random uppercase character.
+    [<CompiledName("CharUpper")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 7]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let upper : Gen<char> =
         char 'A' 'Z'
 
     /// Generates an ASCII character: '\000'..'\127'
+    [<CompiledName("CharAscii")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 8]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let ascii : Gen<char> =
         char '\000' '\127'
 
     /// Generates a Latin-1 character: '\000'..'\255'
+    [<CompiledName("CharLatin1")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 9]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let latin1 : Gen<char> =
         char '\000' '\255'
 
     /// Generates a Unicode character, excluding invalid standalone surrogates:
     /// '\000'..'\65535' (excluding '\55296'..'\57343')
+    [<CompiledName("Char")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Why not just `Unicode`?

File: src/Hedgehog/Gen.fs [hunk 10]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let unicode : Gen<char> =
         filter (not << System.Char.IsSurrogate) unicodeAll
 
     // Generates a random alpha character.
+    [<CompiledName("CharAlpha")>]
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 11]
```diff
@@ -305,63 +342,76 @@ module Gen =
...
     let alpha : Gen<char> =
         choice [lower; upper]
 
+    [<CompiledName("CharAlphaNumeric")>]
     // Generates a random alpha-numeric character.
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we need to prefix this with `Char`?

File: src/Hedgehog/Gen.fs [hunk 12]
```diff
@@ -479,6 +546,7 @@ module Gen =
 
 [<AutoOpen>]
 module GenBuilder =
+    [<CompiledName("Gen")>]
     let gen = Gen.Builder ()
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Do we really need a `[<CompiledName>]` here? Maybe this should be `private`?

File: src/Hedgehog/Range.fs [hunk 13]
```diff
@@ -212,8 +222,9 @@ module Range =
     /// >>> Range.bounds 99 <| Range.linear 0 10
     /// (0, 10)
     ///
-    let inline linear (x : 'a) : ('a -> Range<'a>) =
-      linearFrom x x
+    [<CompiledName("Linear")>]
+    let inline linear (x : 'a) (y : 'a) : Range<'a> =
+      linearFrom x x y
```
Comment (moodmosaic) [inline src/Hedgehog/Range.fs]: What's the motivation for this change? 😃 The idea here is that this, ported, code tries to be as close as possible with the Haskell version; we even ported the doctests 🐻

File: tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/LinqTests.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        //}
+
+        //[Fact]
+        //public void ShrinkTestt()
+        //{
+        //    Check(
+        //        from s in ForAll(Gen.String(Range.Constant(1, 100), Gen.CharAlpha))
+        //        from s2 in ForAll(Gen.String(Range.Constant(1, 10), Gen.CharAlpha))
+        //        select !s.Contains(s2));
+        //}
```
Comment (moodmosaic) [inline tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/LinqTests.cs]: Maybe we should remove this commented-out blocks of code and just add some `FIXME`s?

The problem here is that these might get stale, and after some time we might not know what to do with them; and they might end up staying there, commented-out, forever.

File: tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/NameTests.cs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+        }
+
+        [Theory]
+        [MemberData(nameof(AllPublicMembers))]
+        public void AllPublicMembersFollowDotNetNamingGuidelines(MemberInfo mi)
+        {
+            var startsWithUppercaseChar = char.IsUpper(mi.Name.First());
+            Assert.True(startsWithUppercaseChar, $"{mi.Name} should start with uppercase letter");
+        }
```
Comment (moodmosaic) [inline tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/NameTests.cs]: I think that's pretty neat 💯 It actually looks similar to what AutoFixture.Idioms does 👍

File: src/Hedgehog/LinqSupport.fs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    let select (g : Gen<'a>) (f : Func<'a, 'b>) : Gen<'b> =
+        Gen.map (fun x -> f.Invoke x) g
+
+    [<Extension>]
+    [<CompiledName("SelectMany")>]
+    let bind2 (g : Gen<'a>) (f : Func<'a, Gen<'b>>) (proj : Func<'a, 'b, 'c>) : Gen<'c> =
+        gen {
+            let! a = g
+            let! b = f.Invoke a
+            return proj.Invoke(a, b)
```
Comment (moodmosaic) [inline src/Hedgehog/LinqSupport.fs]: Nitpick: Throughout the F# codebase we add a space between the member/function and its argument. :neckbeard: (near `proj.Invoke`)

File: src/Hedgehog/LinqSupport.fs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+    [<Extension>]
+    [<CompiledName("Select")>]
+    let select (p : Property<'a>) (f : Func<'a, 'b>) : Property<'b> =
+        Property.map (fun x -> f.Invoke x) p
+
+    [<Extension>]
+    [<CompiledName("SelectMany")>]
+    let bind2 (pa : Property<'a>) (f : Func<'a, Property<'b>>) (proj : Func<'a, 'b, 'c>) : Property<'c> =
+        Property.bind pa (fun a -> Property.bind (f.Invoke a) (fun b -> Property.success (proj.Invoke(a, b))))
```
Comment (moodmosaic) [inline src/Hedgehog/LinqSupport.fs]: Nitpick: Throughout the F# codebase we add a space between the member/function and its argument. :neckbeard: (near `proj.Invoke`)

File: src/Hedgehog/Property.fs [hunk 18]
```diff
@@ -270,14 +287,19 @@ module Property =
...
         bindGen (toGen m) (toGen << k) |> ofGen
 
+    [<CompiledName("ForAll")>]
     let forAll (gen : Gen<'a>) (k : 'a -> Property<'b>) : Property<'b> =
         let prepend (x : 'a) =
             bind (counterexample (sprintf "%A" x)) (fun _ -> k x) |> toGen
         Gen.bind gen prepend |> ofGen
 
+    [<CompiledName("ForAll")>]
+    let forAll' (gen : Gen<'a>) : Property<'a> = forAll gen success
```
Comment (moodmosaic) [inline src/Hedgehog/Property.fs]: Nitpick^2:

```f#
let forAll' (gen : Gen<'a>) : Property<'a> = forAll gen success
```

might as well be written as

```f#
let forAll' (gen : Gen<'a>) : Property<'a> =
    forAll gen success
```

so that it's consistent with the rest of the function/member declarations in the codebase. I think it's the same in Haskell version too. /cc @jystic

File: src/Hedgehog/Gen.fs (hunk 12 above)
Comment (Porges) [inline src/Hedgehog/Gen.fs]: Yes I should remove this, leftover from earlier version

File: src/Hedgehog/Range.fs (hunk 13 above)
Comment (Porges) [inline src/Hedgehog/Range.fs]: This is so we don't end up with `FSharpFunc` in the API, otherwise it's tricky for C# to call.

File: src/Hedgehog/Gen.fs (hunk 3 above)
Comment (Porges) [inline src/Hedgehog/Gen.fs]: I'll follow up about the naming later

File: tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/LinqTests.cs (hunk 14 above)
Comment (Porges) [inline tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/LinqTests.cs]: I'm going to turn these into proper tests, just thought I'd get some feedback on the PR while I'm in progress ☺️

File: src/Hedgehog/Gen.fs (hunk 2 above)
Comment (jystic) [inline src/Hedgehog/Gen.fs]: In the Haskell version I actually split this in to two separate combinators [choice](http://hackage.haskell.org/package/hedgehog-0.4.1/docs/Hedgehog-Gen.html#v:choice) + [recursive](http://hackage.haskell.org/package/hedgehog-0.4.1/docs/Hedgehog-Gen.html#v:recursive) which you combine together, so I think this is fine for now.

File: src/Hedgehog/Property.fs (hunk 18 above)
Comment (jystic) [inline src/Hedgehog/Property.fs]: Yep, @moodmosaic is correct regarding the style we use.

File: src/Hedgehog/Range.fs (hunk 13 above)
Comment (jystic) [inline src/Hedgehog/Range.fs]: I think this is a good change, it won't make a difference to the F# usage and the C# usage will be nicer.

File: tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/LinqTests.cs [hunk 19]
```diff
@@ -0,0 +1,126 @@
+﻿using Xunit;
+
+// Import Check and ForAll:
+using static Hedgehog.Property;
```
Comment (jystic) [inline tests/Hedgehog.CSharp.Tests/Hedgehog.CSharp.Tests/LinqTests.cs]: Wow, I've been out of the C# loop for a while, it's awesome that you can do this now 💯

**Comment by jystic** [comment]
I really like the direction here, if you can address @moodmosaic's comments then I'd be more than happy to go ahead with this.

**Comment by jack-pappas** [comment]
@Porges Any update on this? It'd be handy to be able to utilize Hedgehog from C# as well as F#, as one can currently do with [FsCheck](https://github.com/fscheck/FsCheck).

**Comment by Porges** [comment]
@jack-pappas yes I do intend to clean this up soon

**Comment by Porges** [comment]
... now that I'm back from vacation 😁

**Comment by moodmosaic** [comment]
@Porges, if you're busy, that's fine 🎉 🚀 Perhaps I can give it a shot so we can merge it?

**Comment by Porges** [comment]
Updated to fix the stylistic issues. I'm not sure if the build will pass though... I may need to downgrade the .csproj (will see in a moment).

**Comment by Porges** [comment]
Okay, I'll need to downgrade the .csproj. I've been fighting for a while to try and get build.cmd working locally so that I can check the CI build... 🤦‍♂️

There's also something I alluded to in the first post, about not having to write your tests in LINQ syntax, since the body could be quite long (and imperative!), which I'll add before we get this completed. One more day :grin:

File: src/Hedgehog/LinqSupport.fs [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+open System.Runtime.CompilerServices
+
+open Hedgehog
+
+[<Extension>]
+module GenLinqSupport =
+
+    // Support for `where`:
+    [<Extension>]
+    [<CompiledName("Where")>]
```
Comment (moodmosaic) [inline src/Hedgehog/LinqSupport.fs]: Nitpick: This can also be expressed as `[<CompiledName "Where>]` without parenthesis, but I can go ahead and do this after we merge the PR 😅

Comment (Porges) [reply src/Hedgehog/LinqSupport.fs]: That looks a bit weird to me 👽

Comment (moodmosaic) [reply src/Hedgehog/LinqSupport.fs]: Ugh, nevermind then 😅

Comment (Porges) [reply src/Hedgehog/LinqSupport.fs]: happy to change it though, but I haven't seen other F# code doing that... 🙂

Comment (moodmosaic) [reply src/Hedgehog/LinqSupport.fs]: No, don't let this small thing slow you down 😃 

---

It's just that, the closest to our existing code style would really be

```f#
[<CompiledName ("Where")>] // Notice the space.  But then, the space isn't needed.
```

**Comment by Porges** [comment]
I've added a `CheckForAll` method so that you can write bigger properties from C# than is possible in Linq syntax.

It does feel a little clunky, it would be better if there were just a single method that took one `Gen` and if you wanted multiple, then you would pass `Gen.Tuple(Gen.One, Gen.Two, Gen.Three)`... but C# doesn't have destructuring of tuples in lambda arguments, so you'd end up with:

    CheckForAll(
        Gen.Tuple(Gen.One, Gen.Two, Gen.Three),
        tuple => {
            var (one, two, three) = tuple;
            // proceed with the real test...
        });

... which is a whole different clunkiness.

**Comment by moodmosaic** [comment]
@Porges, we now have [fsharp-hedgehog-experimental](https://github.com/hedgehogqa/fsharp-hedgehog-experimental) for trying out new things, like those in https://github.com/hedgehogqa/fsharp-hedgehog/pull/113/commits/ed75041fe59082153a0a659128c41cfabee2db10 😃

The build still fails, so I can't test it locally 😕

**Comment by moodmosaic** [comment]
I mean, it'd be great if we can focus just on the essential stuff to enable support for C#/LINQ and we can tackle all the rest separately, here or in [fsharp-hedgehog-experimental](https://github.com/hedgehogqa/fsharp-hedgehog-experimental).

**Comment by Porges** [comment]
Putting it into `experimental` is a good idea, it would be nice to get some real-world usage to figure out what would work best. It does mean you have to put the main body of the test into another method or a local method if you want to write multiple lines of test, but that's not _too_ onerous 😉 

I've backed out `CheckForAll` for now and downgraded the test project, hopefully this fixes it 🙂

**Comment by moodmosaic** [comment]
Weird, the build server is green but the build fails locally (Visual Studio 2015, F# 4.0):

```
Build FAILED.

       "E:\Notes\fsharp-hedgehog\Hedgehog.sln" (Rebuild target) (1) ->
       "E:\Notes\fsharp-hedgehog\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Te
       sts.csproj" (Rebuild target) (4) ->
       (CoreCompile target) ->
         NameTests.cs(32,48): error CS1026: ) expected [E:\Notes\fsharp-hedgeho
       g\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Tests.csproj]
         NameTests.cs(32,68): error CS1002: ; expected [E:\Notes\fsharp-hedgeho
       g\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Tests.csproj]
         NameTests.cs(32,68): error CS1513: } expected [E:\Notes\fsharp-hedgeho
       g\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Tests.csproj]
         NameTests.cs(37,46): error CS1026: ) expected [E:\Notes\fsharp-hedgeho
       g\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Tests.csproj]
         NameTests.cs(38,90): error CS1002: ; expected [E:\Notes\fsharp-hedgeho
       g\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Tests.csproj]
         NameTests.cs(38,90): error CS1513: } expected [E:\Notes\fsharp-hedgeho
       g\tests\Hedgehog.CSharp.Tests\Hedgehog.CSharp.Tests.csproj]

    0 Warning(s)
    6 Error(s)
```

**Comment by moodmosaic** [comment]
Hmm, it's because I don't have the most recent version of the C# compiler installed, I guess.

**Comment by moodmosaic** [comment]
According to [this](https://stackoverflow.com/a/43048660/467754) answer, here's what needs to be done to support also VS2015:

```diff
$ git diff
diff --git a/paket.dependencies b/paket.dependencies
index fe00929..2a46851 100644
--- a/paket.dependencies
+++ b/paket.dependencies
@@ -10,6 +10,7 @@ source https://www.nuget.org/api/v2/
 // So, we specify 4.0.0.1 as a minimum bound, and set lowest_matching
 // so it won't be updated automatically to a newer one by 'paket update'.
 nuget FSharp.Core >= 4.0.0.1 lowest_matching: true
+nuget Microsoft.Net.Compilers 2.4.0

 nuget FAKE 4.41.1
 nuget xunit.runner.console 2.1.0
diff --git a/tests/Hedgehog.CSharp.Tests/paket.references b/tests/Hedgehog.CSharp.Tests/paket.references
index b023a6b..e355b18 100644
--- a/tests/Hedgehog.CSharp.Tests/paket.references
+++ b/tests/Hedgehog.CSharp.Tests/paket.references
@@ -3,3 +3,4 @@ xunit.core
 xunit.runner.console
 xunit.runner.visualstudio
 FSharp.Core
+Microsoft.Net.Compilers
\ No newline at end of file
```

File: tests/Hedgehog.CSharp.Tests/NameTests.cs [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+                BindingFlags.Static |
+                BindingFlags.DeclaredOnly;
+
+            foreach (var type in _publicApiTypes)
+            {
+                var members = type.GetMembers(bindingFlags);
+
+                foreach (var member in members)
+                {
+                    // Ignore special members like indexers etc.
```
Comment (moodmosaic) [inline tests/Hedgehog.CSharp.Tests/NameTests.cs]: The body the `foreach` loop requires features not found in the C# compiler that ships with Visual Studio 2015. We could either refactor the body of this loop to plain-old C# or do something like [this](https://github.com/hedgehogqa/fsharp-hedgehog/pull/113#issuecomment-341953738), so that the project compiles.

Comment (Porges) [reply tests/Hedgehog.CSharp.Tests/NameTests.cs]: Changed this to C#6-compatible code and forced the project to only use C#6.

Comment (moodmosaic) [reply tests/Hedgehog.CSharp.Tests/NameTests.cs]: @Porges, @jystic runs into this on his local:

```
LinqTests.cs(4,23): error CS0305: Using the generic type `Hedgehog.Property<a>' requires `1' type argument(s)
```

It's basically that `using static Hedgehog.Property;` one we were talking about yesterday. You were right that we have to get rid of it (it *does* compile in VS2015 but has issue with Mono)...

Comment (moodmosaic) [reply tests/Hedgehog.CSharp.Tests/NameTests.cs]: I guess, we should probably leave this as-is, and have @jystic install C# 6 on his machine. I think it's fair to have C# 6 as base...

Comment (Porges) [reply tests/Hedgehog.CSharp.Tests/NameTests.cs]: C#6 is 2015 so yes 😁
